[CQT-84] Add reset instruction (#54)
QuTech-Delft/cQASM-spec: docs/language_specification/statements/reset_instruction.md

> A **`reset`** instruction performs a measurement plus a Pauli X on the qubit argument.

Reviewer: ```suggestion
A **`reset`** instruction resets the state of the qubit to $|0\rangle$. It does this by first measuring the qubit and, conditioned on the outcome being 1, applies a Pauli X gate.
```

Author: Done.

> The general form of a reset instruction is as follows:

Reviewer 2: Specify what a reset does at a higher level? i.e it places the qubit to a ground or 0 state. This is done by performing a measurement and a potential Pauli X on the qubit argument.

Reviewer 2: Also a reset does not always need to have a Pauli X

Author: @juanboschero Thanks for the comments! Can you please comment as well on the [Confluence](https://qutech-sd.atlassian.net/wiki/spaces/QW/pages/3523051530/Research+reset+instruction+semantic) page dedicated to the investigation of the `reset` instruction?

I think the fact that we don't talk about a conditional Pauli X here is because we don't have yet conditional operations on the language.

Reviewer: In the spec we need to describe what from a cQASM point of view is the meaning of the `reset` instruction. Ultimately, it resets the state of the qubit to the |0> state (in the Z-basis). Then we choose to define further what according to us this means through the implementation definition that IBM uses, _i.e._, the qubit is measured and based on the measurement outcome a Pauli-X is applied when 1, and nothing is done when 0, ultimately resulting in a qubit in the |0> state.

So if backends support the `reset` instruction natively, then they need to implement it accordingly. If they instead reset the state of the qubit in another way, they need to make sure that the result is the same as measuring it and performing a conditional Pauli-X gate, otherwise it doesn't perform a `reset` instruction as cQASM users expect it.

So it is important that the 'meaning' of the `reset` instruction is given in the spec.

Note that, because we currently do not support conditional statements in the language (yet), only backends that support the `reset` instruction natively can make use of it, _i.e._, we cannot yet decompose the `reset` instruction in OpenSquirrel, as the language is not rich enough, atm.

Author: OK, can you please then add a suggestion for the `reset` instruction text?

But notice that these are the current texts for both `measure` and `reset` instructions:

> A **`measure`** instruction performs a measurement to its qubit argument and
assigns the outcome to a bit variable.
> A **`reset`** instruction performs a measurement plus a Pauli X on the qubit argument.

    &emsp; __`reset`__ _qubit-argument_

    _qubit-argument_:  
    &emsp; _qubit-variable_  
    &emsp; _qubit-index_

    _qubit-variable_:  
    &emsp; _identifier_

    _qubit-index_:  
    &emsp; _index_

!!! example
    
    === "Reset of all the qubits"
    
        ```linenums="1", hl_lines="3"
        qubit q
        qubit[5] qq
        reset
        ```
    
    === "Reset of a single qubit"
    
        ```linenums="1", hl_lines="2"
        qubit q
        reset q
        ```
    
    === "Reset of multiple qubits through their register index"
    
        ```linenums="1", hl_lines="2"
        qubit[5] q
        reset q[2, 4]
        ```

!!! note

    The reset instruction accepts [SGMQ notation](gates.md#single-gate-multi-qubit-sgmq-notation), similar to gates.

The following code snippet shows how the reset instruction might be used in context.

```linenums="1" hl_lines="8"
version 3.0

qubit[2] q

H q[0]
CNOT q[0], q[1]

reset q  // Measurement in the standard basis, plus a Pauli X on q
```

On the last line of this simple cQASM program,
the respective states of both qubits in the qubit register are measured along the standard/computational basis.

> Then a Pauli X is performed on both qubits.

Author: OK, but I'm then adding as well a measure instruction at the end of the program. Because I see that, what you are trying to explain here, is how `reset` interacts with `measure`.

Also, I wouldn't use terms such as _big-endian_ in this documentation. I would just say, for example for a string such as 01101100, if the lowest qubit indices go at the rightest or leftest positions.

> Then a Pauli X is performed on both qubits.

Author: Also, I wouldn't use terms such as _big-endian_ in this documentation. I would just say, for example for a string such as 01101100, that the lower qubit indices go at the right and the higher at the left.

This note could also be added on the General Overview page, in a section similar to _About the grammar sections_ (something like _About the state strings_). Maybe renaming them to: _How to read the grammar sections_, and _How to read the state strings_.

> Then a Pauli X is performed on both qubits.

Author: Done.

!!! example
    
    === "Reset of all the qubits"
    
        ```linenums="1", hl_lines="3"
        qubit q
        qubit[5] qq
        reset
        ```
    
    === "Reset of a single qubit"
    
        ```linenums="1", hl_lines="2"
        qubit q
        reset q
        ```
    
    === "Reset of multiple qubits through their register index"
    
        ```linenums="1", hl_lines="2"
        qubit[5] q
        reset q[2, 4]
        ```

!!! note

    The reset instruction accepts [SGMQ notation](gates.md#single-gate-multi-qubit-sgmq-notation), similar to gates.

The following code snippet shows how the reset instruction might be used in context.

```linenums="1" hl_lines="9"
version 3.0

qubit[2] q
bit[2] b

H q[0]
CNOT q[0], q[1]

reset q[0]  // Resets the state of qubit q[0] to |0>

b = measure q
```

The `reset` instruction is performed by measuring `q[0]` along the computational basis.
Based on the measurement outcome, either no operation is performed (in case the outcome is 0) or
a Pauli X gate is applied (in case the outcome is 1).

The `measure` instruction should report $|00\rangle$ in 50% of the cases and $|10\rangle$ in the other 50% of the cases

> (where lower qubit indices are shown here at the right of the string and higher indices at the left).

Author: Done. Except for that `(line 11)`. It doesn't add much, and it's difficult to maintain.

    &emsp; _qubit-index_

    _qubit-variable_:  
    &emsp; _identifier_

    _qubit-index_:  
    &emsp; _index_

!!! example
    
    === "Reset of all the qubits"
    
        ```linenums="1", hl_lines="3"
        qubit q
        qubit[5] qq
        reset
        ```
    
    === "Reset of a single qubit"
    
        ```linenums="1", hl_lines="2"
        qubit q
        reset q
        ```
    
    === "Reset of multiple qubits through their register index"
    
        ```linenums="1", hl_lines="2"
        qubit[5] q
        reset q[2, 4]
        ```

!!! note

    The reset instruction accepts [SGMQ notation](gates.md#single-gate-multi-qubit-sgmq-notation), similar to gates.

The following code snippet shows how the reset instruction might be used in context.

```linenums="1" hl_lines="9"
version 3.0

qubit[2] q
bit[2] b

H q[0]
CNOT q[0], q[1]

reset q[0]  // Resets the state of qubit q[0] to |0>

b = measure q
```

The `reset` instruction is performed by measuring `q[0]` along the computational basis.
Based on the measurement outcome, either no operation is performed (in case the outcome is 0) or
a Pauli X gate is applied (in case the outcome is 1).

> 

Author: Can you please add this as a suggestion?

> 

Author: ```suggestion
```

A **`reset`** instruction resets the state of the qubit to $|0\rangle$.
It does this by first measuring the qubit and then, conditioned on the outcome being 1, applying a Pauli X gate.

!!! note
    

>     Even though, the `reset` instruction internally measures the qubit state,

Author: ```suggestion
    Even though the `reset` instruction internally measures the qubit state,
```

    it does not store the measurement outcome in the measurement register,

>     _i.e._ the measurement register is unaffected by the `reset` instruction.

Author: ```suggestion
    _i.e._, the measurement register is unaffected by the `reset` instruction.
```
